### DISCUSSION TITLE
Subscribing to multiple brokers

### DISCUSSION THREAD

**Discussion by sussex1** [bellrichm/WeeWX-MQTTSubscribe#138] [Q&A]
Hi
Is it possible use multiple instances  to subscribe to multiple MQTT brokers or should I aggregate data elsewhere?
;-)

**Comment by bellrichm** [comment]
If by 'multiple instances', you mean running multiple WeeWX instances, I would think that would work.
MQTTSubscribe currently only supports connecting to a single broker. I've thought about supporting multiple brokers, but haven't heard any real demand.

**Reply by simplegui-helper** [reply]
It sure would be nice.  One broker on the Internet for my public Belchertown skin and a private broker for the LAN for by a Belchertown kiosk.  My plan is to have a kiosk displaying 24/7 on a 10" Raspberry Pi in my kitchen.  If I had a private LAN broker, I would not need access to the Internet to run the Belchertown kiosk.  How much work would it be to allow multiple MQTT brokers?

**Reply by bellrichm** [reply]
@simplegui-helper 
If I understand you, It sounds like you want to be able to publish to multiple brokers. And have different weewx instances subscribe (via Belchertown skin). 
If this is what you want, that would be in the code you are using to publish (probably weewx-mqtt…).
If I am not understanding you, could you proved some more information?

**Reply by simplegui-helper** [reply]
Yes, this is what I was asking.  Since my post, I have done the following...

Copied mqtt.py to mqtt2.py and changed all names in mqtt2.py from MQTT to MQTT2.  Added MQTT2 to the weewx.config file.  I installed and setup a new broker on my local LAN weewx system.  It is now publishing to both MQTT (cloud server) and MQTT2 (local LAN)!  I have hacked the belchertown.js.tmpl and skin.config files so that index.html uses the MQTT broker and the kiosk.html uses the MQTT2 broker with different credentials (added "mqtt_websockets_username_kiosk" and "mqtt_websockets_password_kiosk").

It is working!

I also considered using a local MQTT broker to bridge to the MQTT broker in the cloud.  I didn't do it because I could not tailor the packets for just my local LAN kiosk system which are different that what I publish to the cloud.  You can see my web page at http://rr3weather.com.

**Reply by bellrichm** [reply]
@simplegui-helper 
Sounds good.  Thanks for the update.

**Reply by clydebarrow** [reply]
> Copied mqtt.py to mqtt2.py and changed all names in mqtt2.py from MQTT to MQTT2. Added MQTT2 to the weewx.config file.

I've done that, but the mqtt2.py is not being invoked. What else did you have to do to make weewx aware of it?

Ah, got it - had to add user.mqtt2.MQTT2 to Engine/Services/restful_services in weewx.conf.

**Comment by peters77** [comment]
It would be really nice to get data from 2 MQTT brokers. One is my personal MQTT Server at home feeded by a LoRa sensor. This setup is now running. (Thanks Rich!). 

```
[MQTTSubscribeService1]
    host = 192.168.1.100

    #    console = false
    #    log = false
    #    [[message_callback]]
    #        type = json
    [[topics]]
        ignore_start_time = True
        [[[weather]]]  ....
```

I have a second LoRa sensor placed far away which has only connection to a TTN gateway 3km far away. The TTN service provides  a MQTT server, too. I checked and I can subscribe via mosquitto_sub.

```
[MQTTSubscribeService2]
    host = eu1.cloud.thethings.network
    port = 8883 ....

```
I read that some people were able to start a second instance of WeeWX-MQTTSubscribe? to connect to a second MQTT server in parallel inside weewx. 
Is there a wiki or howto to setup a second MQTTSubscribe service and howto change the code for the second instance/.py file? 
As far as I understnad this is not possible at the moment to provide two different MQTT adresses and topics list with one instance of MQTTSubscribe running!?  But I fear it's to complicated and needs lots of code change....not for end users doable....? :joy:

Regards,

Christian

**Reply by bellrichm** [reply]
Currently subscribing to multiple brokers is not supported. You could probably make a copy of MQTTSubscribe.py and hack it to have a different configuration stanza, like some did for the MQTT publish extension.
I did open [226](https://github.com/bellrichm/WeeWX-MQTTSubscribe/issues/226) to fully support this. But, at this time it is pretty low on my list of things to do… Mostly because it looks like it would be a major breaking change to the configuration format. I’d also want to delve into the possible performance implications of subscribing to multiple brokers.

**Reply by peters77** [reply]
Yes, I think hacking into it like one user did with the MQTT publish extension is about my horizon of coding skills. But thanks for open https://github.com/bellrichm/WeeWX-MQTTSubscribe/issues/226. And yes, could be timing problem to subscribe to different brokers which different latency.

**Comment by n7qnm** [comment]
It would be a bit of kludge; but you could write a script that would subscribe to the cloud publisher and publish to yours.

 

Clay

N7QNM

**Comment by peters77** [comment]
Yes, I thought about this and search the web and I wonder that nobody seems to do that so far?! Can't find anything...susbcribing to TTN, read out my values and publish it to my local mosquitto server would be a solution...! 👍  I will have a look....but my coding skills are low.... 😄

**Comment by earlthesquirrel** [comment]
I have written one for just this purpose.   I have some devices that publish and I can’t really control what topics they publish to, so I wrote something to listen to those and also other topics, etc.   I can share it, and/or contribute it to this package ( another happy customer of this skin )

**Reply by peters77** [reply]
That would be great if you could share it, would give it a try.   :slightly_smiling_face:

**Reply by earlthesquirrel** [reply]
[mqttReflector-simple.txt](https://github.com/user-attachments/files/18539709/mqttReflector-simple.txt)

Added as a text file (should be a .py file, but GitHub complained)

This is an edited down version of what I use.  I re-direct by topic... mine has a bunch more -- I redirect from 5 different devices that publish on their own topics.  Instead of changing the firmware on them, I just use this and add the topic they post on to the file.  I also drop data that really isn't important to what I'm recording (like RSSI).    Hopefully this is of some assistance.

BTW, the IP address in the file should point to the server you want to post to. I left in the non-routable address for my local server in there.  The user id and password is for that server as well.  Right now, I'm only using a single server, (but lots of different topics) but I had in the past used multiple...it should be fairly simple to recreate that part.

**Reply by peters77** [reply]
Thank you! Will give it a try. I will try to tweak it for my setup! 👍

**Comment by peters77** [comment]
As my python skills are very limited I found another solution which I would share: 
mosquitto has a build in bridge mode. So I can subscribe to a remote broker and push the messages to my local mosquitto instance. 
I downloaded the CA file from TTN and added for TTN in `/etc/mosquitto/mosquitto.conf `


```
# Bridge to The Things Network
connection bridge
bridge_protocol_version mqttv311
notifications false
cleansession true
address eu1.cloud.thethings.network:8883
remote_username xyz@ttn         # username ttn
remote_password password       # API Key from TTN
bridge_cafile /etc/mosquitto/certs/cacert.pem
bridge_insecure false
try_private false
topic # both 2 ttn/ ""

```
Now the remote data is published at my local mosquitto instance in /ttn/#.
Works good for now. 👍

**Reply by bellrichm** [reply]
Cool. I was wondering about bridge mode. Just haven’t found time to check it out.
